### PR TITLE
"did you mean" for ObjectParser with top named

### PR DESCRIPTION
**Pull request by nik9000**
When you declare an ObjectParser with top level named objects like we do
with `significant_terms` we didn't support "did you mean". This fixes
that.

Relates #50938

### CODE REVIEW

File: libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java [hunk 1]
```diff
@@ -123,19 +123,18 @@ public NamedXContentRegistry(List<Entry> entries) {
...
         if (entry == null) {
-            throw new NamedObjectNotFoundException(parser.getTokenLocation(), "unable to parse " + categoryClass.getSimpleName() +
-                " with name [" + name + "]: parser not found");
+            throw new NamedObjectNotFoundException(parser.getTokenLocation(), "unknown field [" + name + "]", parsers.keySet());
         }
         if (false == entry.name.match(name, parser.getDeprecationHandler())) {
             /* Note that this shouldn't happen because we already looked up the entry using the names but we need to call `match` anyway
              * because it is responsible for logging deprecation warnings. */
-            throw new NamedObjectNotFoundException(parser.getTokenLocation(),
+            throw new XContentParseException(parser.getTokenLocation(),
```
Comment (nik9000) [inline libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java]: All of the exceptions that I changed to `XContentParseException`  are really more development errors than usage errors.

Comment (nik9000) [reply libs/x-content/src/main/java/org/elasticsearch/common/xcontent/NamedXContentRegistry.java]: And they don't have the critical `candidates` set.

**Review by romseygeek — APPROVED** [review]
LGTM.  Looking at other places that NamedObjectNotFoundException is used in the code, does this mean that we can rewrite `AbstractQueryBuilder.parseInnerQueryBuilder` to use an ObjectParser and get 'did you mean' suggestions for misspelt queries?

**Comment by nik9000** [comment]
> Looking at other places that NamedObjectNotFoundException is used in the code, does this mean that we can rewrite `AbstractQueryBuilder.parseInnerQueryBuilder` to use an ObjectParser and get 'did you mean' suggestions for misspelt queries?

I believe so.

**Comment by nik9000** [comment]
> I believe so.

I think we'd more likely want to port the `SearchSourceBuilder`'s parsers to `ObjectParser` and have the `query` field use named x content. So long as that works we can avoid needing that method entirely.

**Pull request merged**